### PR TITLE
[air] wip - consolidate train/tune report/checkpoint APIs.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

https://docs.google.com/document/d/1r4LuEVaqFwG5UHREwo34kdgTgno-ng6qK_PC_XV5jpE/edit#

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/ml/session.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+_DATA_DICT = "data_dict"
+
+
+def _get_session():
+    # TODO(xwjiang): Currently the ml session only multiplexes between Tune and Train session.
+    #  We should extract common functionalities if possible.
+    from ray.tune.session import _session as tune_session
+    from ray.train.session import _session as train_session
+    if train_session and tune_session:
+        logger.warning("Expected to be either in tune session or train session but not both.")
```
Comment (krfricke) [inline python/ray/ml/session.py:24]: I'm wondering if we're actually always in a train+tune session if using train in tune

File: python/ray/ml/session.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        return None
+    if not (train_session or tune_session):
+        logger.warning("In neither tune session nor train session!")
+        return None
+    return train_session or tune_session
+
+
+def save(epoch: int, checkpoint: Optional[Checkpoint] = None, metrics: Optional[dict] = None):
```
Comment (krfricke) [inline python/ray/ml/session.py:32]: Let's have a chat with some people about APIs here! I think we should usually use `save_checkpoint` if we're saving checkpoints here, except if we use the API to save other things as well, like artifacts.

Another option would be to consolidate `session.report` with this - e.g. `session.report(some_metric=a, checkpoint=cp)` or so.

I think a mini design doc would be great here!

File: python/ray/ml/session.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            session.save(epoch=0, checkpoint=checkpoint)
+
+        tuner = Tuner(trainable=training_func)
+        tuner.fit()
+
+    """
+    is_dict = None
+    if checkpoint:
+        # This checkpoint can be dict checkpoint or a directory checkpoint.
+        if not checkpoint.get_internal_representation()[0] in (_LOCAL_PATH, _DATA_DICT):
```
Comment (krfricke) [inline python/ray/ml/session.py:74]: Can't we support all kinds of checkpoints here?

Comment (krfricke) [inline python/ray/ml/session.py:74]: Can't we support all kinds of checkpoints here?

Comment (krfricke) [reply python/ray/ml/session.py:74]: E.g. an object ref checkpoint is basically a dict checkpoint, a URI checkpoint is basically a local path checkpoint. I'm thinking about distributed training cases

**Review by krfricke — COMMENTED** [review]
Let's discuss the API, but I think the multiplexing is a great start!

File: python/ray/ml/session.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/ml/session.py:24]: shouldn't happen. Trainable spawns Train workers, which don't have access to the tune session.

**Pull request closed without merging**